content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,386 @@
+    Compute the RFFT of the real-valued array T using pyFFTW.
+    """
+    a = pyfftw.empty_aligned(len(T), dtype='float64')
+    a[:] = T
+    rfft_object = pyfftw.builders.rfft(a)
+    R = rfft_object()
+
+    return R
+```
+
+### First Attempt: Reuse array (v1)
+One potential optimization is to avoid creating a byte-aligned array on every call. Instead, we can create the byte-aligned array once and reuse it for subsequent calls as long as the size of the input array does not change. This way we are saving the overhead of memory allocation on every call.
+
+```python
+class rfft_caller_v1:
+    def __init__(self):
+        self.real_arr = None
+
+    def __call__(self, T):
+        if self.real_arr is None or len(T) != len(self.real_arr):
+            self.real_arr = pyfftw.empty_aligned(len(T), dtype='float64')
+        
+        rfft_obj = pyfftw.builders.rfft(self.real_arr)
+        self.real_arr[:] = T
+        rfft_obj.execute()
+        
+        return rfft_obj.output_array
+
+_rfft_caller_v1 = rfft_caller_v1()
+```
+
+In this modified version, an instance of the class `rfft_caller_v1` can be used to compute the RFFT. It holds the byte-aligned array `real_arr`, as an attribute, after the first call. When the `__call__` method is invoked again, it checks if the size of the input array `T` matches the size of the existing aligned array. If yes, it reuses the existing byte-aligned array. Let's update the dictionary `rfft_callers` to include this new version:
+
+```python
+rfft_callers = {
+    'v0': _rfft_caller_v0,
+    'v1': _rfft_caller_v1,
+}
+```
+
+Let's check out the performance improvement. The blue line higher than the red baseline implies "faster". 
+
+![Performance Gain](Figure_v1.png)
+
+As shown in the performance plot above, our first attempt to optimize the RFFT computation resulted in about 2-12% speed-up for most input sizes. It is not clear to me why the performance drops for inputs with sizes `2^13` to `2^15`. I am going to skip the investigation for those cases. Let's see if we can do better!!
+
+
+### Second Attempt: Tell RFFT to avoid copy (v2)
+According to [pyfftw.builders.rfft documentation](https://pyfftw.readthedocs.io/en/latest/source/pyfftw/builders/builders.html#module-pyfftw.builders): 
+
+> By default, these functions will always create a copy (and sometimes more than one) of the passed-in input array. This is because the creation of the pyfftw.FFTW object generally destroys the contents of the input array.
+
+> This only influences a copy during the creation of the object
+
+In our case, the byte-aligned array `real_arr` is used when creating the RFFT object. And, then, it is filled with the contents of `T` right before executing the RFFT object. Therefore, at the time of creating the RFFT object, the contents of `real_arr` are not important as they will be overwritten. We do this by setting the `avoid_copy` parameter to `True` to avoid unnecessary copying of data at the time of creating the RFFT object.
```

Comment:
> We do this by setting the `avoid_copy` parameter to `True` to avoid unnecessary copying of data at the time of creating the RFFT object.


Therefore, we can set the `avoid_copy` parameter to `True` to avoid unnecessary copying of data at the time of creating the RFFT object.